e2e_playwright/st_file_uploader_test.py

```diff
@@ -0,0 +1,218 @@
+    )
+
+    assert_snapshot(file_uploader_in_error_state, name="st_file_uploader-error")
+
+
+def test_uploads_and_deletes_single_file_only(
+    themed_app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Test that uploading a file for single file uploader works as expected."""
+    file_name1 = "file1.txt"
+    file_content1 = b"file1content"
+
+    file_name2 = "file2.txt"
+    file_content2 = b"file2content"
+
+    uploader_index = 0
+
+    with themed_app.expect_file_chooser() as fc_info:
+        themed_app.get_by_test_id("stFileUploadDropzone").nth(uploader_index).click()
+
+    file_chooser = fc_info.value
+    file_chooser.set_files(
+        files=[{"name": file_name1, "mimeType": "text/plain", "buffer": file_content1}]
+    )
+    wait_for_app_run(themed_app)
+    themed_app.wait_for_timeout(1000)
+
+    expect(themed_app.locator(".uploadedFileName")).to_have_text(
+        file_name1, use_inner_text=True
+    )
+
+    expect(themed_app.get_by_test_id("stText").nth(uploader_index)).to_have_text(
+        str(file_content1), use_inner_text=True
+    )
+
+    file_uploader_uploaded_state = themed_app.get_by_test_id("stFileUploader").nth(
+        uploader_index
+    )
+
+    assert_snapshot(
+        file_uploader_uploaded_state, name="st_single_file_uploader-uploaded"
+    )
+
+    expect(
+        themed_app.get_by_test_id("stMarkdownContainer").nth(uploader_index + 1)
+    ).to_have_text("True", use_inner_text=True)
+
+    # Upload a second file. This one will replace the first.
+    with themed_app.expect_file_chooser() as fc_info:
+        themed_app.get_by_test_id("stFileUploadDropzone").nth(uploader_index).click()
+
+    file_chooser = fc_info.value
+    file_chooser.set_files(
+        files=[{"name": file_name2, "mimeType": "text/plain", "buffer": file_content2}]
+    )
+
+    wait_for_app_run(themed_app)
+    themed_app.wait_for_timeout(1000)
+
+    expect(themed_app.locator(".uploadedFileName")).to_have_text(
+        file_name2, use_inner_text=True
+    )
+
+    expect(themed_app.get_by_test_id("stText").nth(uploader_index)).to_have_text(
+        str(file_content2), use_inner_text=True
+    )
+
+    expect(
+        themed_app.get_by_test_id("stMarkdownContainer").nth(uploader_index + 1)
+    ).to_have_text("True", use_inner_text=True)
+
+    themed_app.get_by_test_id("stHeader").press("r")
+
+    wait_for_app_run(themed_app)
+    themed_app.wait_for_timeout(1000)
+
+    expect(themed_app.get_by_test_id("stText").nth(uploader_index)).to_have_text(
+        str(file_content2), use_inner_text=True
+    )
+
+    themed_app.get_by_test_id("fileDeleteBtn").nth(uploader_index).click()
+
+    wait_for_app_run(themed_app)
+    themed_app.wait_for_timeout(1000)
+
+    expect(themed_app.get_by_test_id("stText").nth(uploader_index)).to_have_text(
+        "No upload", use_inner_text=True
+    )
+
+
+def test_uploads_and_deletes_multiple_files(
+    themed_app: Page, assert_snapshot: ImageCompareFunction
```

Comment:
nit: I usually only run the primary snapshot test via `themed_app` (in this case `test_file_uploader_render_correctly`) and the rest via just `app` to reduce snapshots and runtime. It make also makes sense here to run this with `themed_app`, but only if there are themed styling aspects that are not covered by the other test. Same also with the other test functions.

```suggestion
    app: Page, assert_snapshot: ImageCompareFunction
```